### PR TITLE
Add libsodium deps to Vim/MacVim and fix misc formula issues

### PR DESCRIPTION
**Pull request by ychin**
macvim: Add libsodium support and use build-time-only deps

Vim added libsodium a while ago for handling reading/writing encrypted
files, and we should support it in the Homebrew distribution for MacVim.
Also added checks to make sure `+sodium` is in the version output.

Also, MacVim has been building with statically linked libraries (gettext
and sodium) in its build system, unlike Vim which uses dynamic linking
in its configure script. Because of that, we don't actually have a
runtime dependency on those libraries. Move them to build-time
dependnecy instead.

---

vim: Add libsodium support and fix `brew test vim`

Vim added libsodium a while ago for handling reading/writing encrypted
files, and we should support it in the Homebrew distribution. Also added
checks to make sure `+sodium` is in the version output.

Also, fix `brew test vim` not working on a local machine, since it
cannot find gettext properly. To properly instruct configure to look
under the prefix (e.g. /opt/homebrew), we actually need to pass
`--with-local-dir` to the configure script, as otherwise it's going to
use /usr/local. I think this may be why it's working in build systems
but not my local machine. Note that macvim.rb is already setting this
flag in the configure script and therefore works fine.

---

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by ychin** [comment]
> Also, MacVim has been building with statically linked libraries (gettext and sodium) in its build system, unlike Vim which uses dynamic linking in its configure script

Just to show that this is the case, this is what `otool` reports when I query runtime dependencies on Vim vs MacVim after installing them by using `--build-from-source`:

```
➜ ychin homebrew/Library/Taps/homebrew/homebrew-core/Formula% otool -L /opt/homebrew/Cellar/vim/9.0.1400/bin/vim | grep /opt/homebrew                         
/opt/homebrew/Cellar/vim/9.0.1400/bin/vim:
	/opt/homebrew/opt/ncurses/lib/libncursesw.6.dylib (compatibility version 6.0.0, current version 6.0.0)
	/opt/homebrew/opt/libsodium/lib/libsodium.23.dylib (compatibility version 27.0.0, current version 27.0.0)
	/opt/homebrew/opt/lua/lib/liblua.dylib (compatibility version 5.4.0, current version 5.4.4)
	/opt/homebrew/opt/perl/lib/perl5/5.36/darwin-thread-multi-2level/CORE/libperl.dylib (compatibility version 5.36.0, current version 5.36.0)
	/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/Python (compatibility version 3.11.0, current version 3.11.0)
	/opt/homebrew/opt/ruby/lib/libruby.3.2.dylib (compatibility version 3.2.0, current version 3.2.1)

➜ ychin homebrew/Library/Taps/homebrew/homebrew-core/Formula% otool -L /opt/homebrew/Cellar/macvim/9.0.1276/MacVim.app/Contents/MacOS/Vim | grep /opt/homebrew
/opt/homebrew/Cellar/macvim/9.0.1276/MacVim.app/Contents/MacOS/Vim:
	/opt/homebrew/opt/lua/lib/liblua.dylib (compatibility version 5.4.0, current version 5.4.4)
	/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/Python (compatibility version 3.11.0, current version 3.11.0)
	/opt/homebrew/opt/ruby/lib/libruby.3.2.dylib (compatibility version 3.2.0, current version 3.2.0)
```

**Comment by carlocab** [comment]
> Also, fix `brew test vim` not working on a local machine, since it
> 
> cannot find gettext properly. To properly instruct configure to look
> 
> under the prefix (e.g. /opt/homebrew), we actually need to pass
> 
> `--with-local-dir` to the configure script, as otherwise it's going to
> 
> use /usr/local. I think this may be why it's working in build systems
> 
> but not my local machine.

I don't understand this. `gettext` is also in `/opt/homebrew` on the build machine, so why isn't `--with-local-dir` needed there?

**Comment by ychin** [comment]
> I don't understand this. `gettext` is also in `/opt/homebrew` on the build machine, so why isn't `--with-local-dir` needed there?

Actually, sorry for the misleading statement I wrote. Thinking more about it, does the build machine set `export LDFLAGS="-L$(brew --prefix)/lib"`? I think that's the difference that allows the configure to discover the gettext libs in build systems, but not on a user's machine. If vim.rb manually sets `LDFLAGS` that will work too, but I think using `--with-local-dir` may be cleaner.

**Comment by carlocab** [comment]
> Actually, sorry for the misleading statement I wrote. Thinking more about it, does the build machine set `export LDFLAGS="-L$(brew --prefix)/lib"`?

No. The environment on the build machine will be the same as the environment on your machine when you do
```
brew install --build-from-source
```

File: Formula/vim.rb [hunk 1]
```diff
@@ -55,6 +56,7 @@ def install
     # Homebrew will use the first suitable Perl & Ruby in your PATH if you
     # build from source. Please don't attempt to hardcode either.
     system "./configure", "--prefix=#{HOMEBREW_PREFIX}",
+                          "--with-local-dir=#{HOMEBREW_PREFIX}",
```
Comment (carlocab) [inline Formula/vim.rb:59]:
```suggestion
```
I'm not convinced this is needed, so I'd rather not add this here.

Comment (ychin) [reply Formula/vim.rb:59]: Ok that's fine. I still think it's likely some linker flags in the environment is allowing it to build successfully somehow, but I don't really care about this issue too much tbh. I'll remove it.

Comment (carlocab) [reply Formula/vim.rb]: `brew` sanitises the build environment, so any linker flags available on the build machine should be available when you build from source locally.

**Review by carlocab — APPROVED** [review]

**Comment by carlocab** [comment]
Thanks!

**Pull request merged**